Added an implementation of a comparable Key Value Pair.

JoshKeegan opened the pull request:
Can be used in data structures to give elements a different priority than their own implementation of IComparable<T> will give (as is the case in most use cases).

When switching my code from using your ConcurrentPriorityQueue library to your new DataStructures one, I needed something like this for my use case where the priority is separate from the element. Since that could be quite common, I thought this might be best included with the library.

dshulepov commented:
Hi
I think you are right about the problem - it became not so convinient to use a ConcurrentPriorityQueue when you have priorities separate from values and a convinient way to compare KeyValuePairs just by key would be nice.
However I think your approach is slightly not what is needed. At first, class is a bit heavier thing than a struct, so it'll take more memory. And secondly, comparing item is the responsibility of Comparer.
So, I think providing a built-in KeyValuePair comparer would be a nicer way to solve the problem. 
Would you be interested in implementing a KeyValuePairComparer instead? I could do it myself, thanks to your idea, but I don't want to "still your glory" :)

JoshKeegan commented:
Hi Denis,
That sounds good, I'll change it tomorrow :)
Cheers,
Josh

> On 12 Nov 2015, at 18:49, Denis Shulepov <notifications@github.com> wrote:
> 
> Hi
> I think you are right about the problem - it became not so convinient to use a ConcurrentPriorityQueue when you have priorities separate from values and a convinient way to compare KeyValuePairs just by key would be nice.
> However I think your approach is slightly not what is needed. At first, class is a bit heavier thing than a struct, so it'll take more memory. And secondly, comparing item is the responsibility of Comparer.
> So, I think providing a built-in KeyValuePair comparer would be a nicer way to solve the problem. 
> Would you be interested in implementing a KeyValuePairComparer instead? I could do it myself, thanks to your idea, but I don't want to "still your glory" :)
> 
> —
> Reply to this email directly or view it on GitHub.
>

JoshKeegan commented:
I've now implemented the KeyValuePairComparer. 
I hadn't noticed before that you were already using a comparer elsewhere, so didn't need to make any modifications to PriorityQueue & SkipList making this  fairly simple change.

However, in order for the data structures to work with types that don't implement IComparable<T> (such as KeyValuePair) but have a custom comparer supplied, I've removed the requirement for T to implement IComparable<T>. Instead, the ctor must now enforce that either a comparer must be supplied, or the type must be comparable (through either IComparable<T> or IComparable).

Also, it would be possible for the ctor to automatically select the KeyValuePairComparer if T is a KeyValuePair and no comparer has been specified, but I decided against it for now so the user must explicitly specify a instance of KeyValuePairComparer to the PriorityQueue ctor.

dshulepov commented:
Thanks a lot for contribution. I think you did a nice job.
I did a bit of minor polishing like added dots in the end of the comments and changes some wording.
I've updated the nuget as well.

JoshKeegan commented:
Thanks for updating the nuget package, I've now replaced my use of your ConcurrentPriorityQueue library with this one & it's working well :)
